### DISCUSSION TITLE
Plugin geoip samples

### DISCUSSION THREAD

**Discussion by selukov** [coredns/coredns#4966] [Q&A] [answered]
I'm trying to use the geoip plugin, but I don't understand how to use the Metadata and geoip plugins syntax, could you add a simple example
I used [geodns](https://github.com/abh/geodns) see example.com.json 
```code
{
    "ttl" : 390,
    "serial" : 1345449135,
    "data" : {
        "" : {
             "ns": [
                 "a.ntpns.org",
                 "b.ntpns.org",
                 "x.example.com"
             ],
             "a" : [
                [
                   "203.17.251.1",
                   "1000"
                ]
             ]
        },
        "ru": {
            "a": [
                [
                    "192.168.1.2",
                    "10"
                ]
            ]
        },
        "se": {
            "a": [
                [
                    "192.168.1.3",
                    "10"
                ]
            ]
        }
    }
}
```

Can I do similar using the geoip plugin?

**Comment by snebel29** [comment]
Hi Selukov,
Coredns geoip plugin, add associated geoip information to the metadata of
requests for other plugins to consume.

To the best of my knowledge, no other plugin capable of doing record
serving such as file has integrated with geoip, and therefore geoip is for
now useful only if you implement your own plugin logic.

I have no keyboard around now, I will take a better look later today to
explore adding integration which was something I already had in mind.

I hope that helps.

Cheers.

**Comment by chrisohaver** [comment, marked as answer]
The syntax for using metadata is usually documented with examples in the README of plugins that can use metadata.

As for built in plugins, the _log_ plugin can write metadata to stdout.  e.g. 
```
. {
  geoip file.db
  metadata
  log . "request from: {/geoip/city/name}"
}
```

One external plugin that comes to mind is [_policy_ aka _firewall_](https://github.com/coredns/policy), with which you can block/ignore/allow requests based on metadata values.  Syntax is slightly different than in log. It uses square brackets `[/geoip/city/name]` instead of braces `{/geoip/city/name}`. There are examples in the README there.

**Reply by arLevi** [reply]
The `log . "request from: {/geoip/city/name}"` could be great for debugging, Why it's not working for me ? I copy-pasted the same example yet, the log row is empty after the "request from"? ...
I also noticed somewhere the format of `{/metadata/geoip/city/name}` yet that also doesn't work ...

**Reply by snebel29** [reply]
Log plugin documents how to access labels, here https://github.com/coredns/coredns/blob/master/plugin/log/README.md#log-format and the geoIP plugin the metadata labels here https://github.com/coredns/coredns/tree/master/plugin/geoip#metadata-labels.

I suggest you double check your geoIP and log plugins are configured correctly to start with.

I hope that helps.

**Reply by arLevi** [reply]
Thanks, my testing was wrong, the `geoip` inside the metadata was empty when I run something like `dig myexample.org @localhost` and when I added a subnet like `dig myexample.org @localhost +subnet=SOME_IP/32` then the log showed up with data ...
